pvlib/clearsky.py

```diff
@@ -597,7 +599,132 @@ def _calc_d(aod700, p):
+
+
+def _shift_from_align(align, window):
+    # development code left here as comments in case we decide to support
+    # align='left', 'right' in the future
+    # number of places to shift to line up pandas.Series.rolling with
+    # align value.
+    # pandas.Series.rolling uses kwarg center with values of True or False
+    # default is False which means right aligned labels by default
+    # center=True aligns to center
+    # Here, calculate shift to align='left' (legacy pvlib behavior)
+    # code here for future capability if align='left', 'right' are of interest
+    # commented out to avoid decreasing test coverage
+    # if align == 'left':
+    #     shift = 1 - window
+    # else:
+    #     shift = 0
+    shift = 0
+    return shift
+
+
+def _get_sample_intervals(times, win_length):
+    """ Calculates time interval and samples per window for Reno-style clear
+    sky detection functions
+    """
+    deltas = np.diff(times.values) / np.timedelta64(1, '60s')
+
+    # determine if we can proceed
+    if times.inferred_freq and len(np.unique(deltas)) == 1:
+        sample_interval = times[1] - times[0]
+        sample_interval = sample_interval.seconds / 60  # in minutes
+        samples_per_window = int(win_length / sample_interval)
+        return sample_interval, samples_per_window
+    else:
+        raise NotImplementedError('algorithm does not yet support unequal '
+                                  'times. consider resampling your data.')
+
+
+def _calc_line_length(data, sample_interval):
+    """ Calculates line length for Reno-style clear sky detection functions.
+    """
+    return np.sum(np.sqrt(np.diff(data)**2 + sample_interval**2))
+
+
+def _calc_slope_max_diff(meas, clear, window, align):
+    """ Calculates maximum difference in slope on rolling windows
+    """
+    def _maxdiff(s):
+        return np.abs(np.diff(s)).max()
+
+    center = align == 'center'
+    irrad_diff = meas - clear
+    return irrad_diff.rolling(window, center=center).apply(_maxdiff)
+
+
+def _clear_sample_index(clear_windows, samples_per_window, align, H):
+    """
+    Returns indices of clear samples in clear windows
+    """
+    # first window is in first column of H. clear_windows is aligned by 'align'
+    # shift clear_windows.index so that first window is in left position, lined
+    # up with column of H containing the indices of the first window
+    # commented if/else block for future align='left', 'right' capability
+    # if align == 'right':
+    #     shift = 1 - samples_per_window
+    # elif align == 'center':
+    #     shift = - (samples_per_window // 2)
+    # else:
+    #     shift = 0
+    shift = -(samples_per_window // 2)
+    idx = clear_windows.shift(shift)
+    # drop rows at the end corresponding to windows past the end of data
+    idx.drop(clear_windows.index[1 - samples_per_window:], inplace=True)
+    idx = idx.astype(bool)  # shift added nan, changed type to object
```

Comment:
Seems like pandas makes this more complicated instead of easier. Is this equivalent to the 3 lines above?

```python
idx = clear_windows.iloc[shift:1-samples_per_window]
```

Reply:
No, because the index isn't shifted. Since window statistics are aligned to center by default, the 'shift' changes alignment to 'left' so that the clear index values can be extracted with the Hankel matrix.  I don't recall the circumstances which changed the type to object from bool, but it came up.

Reply:
I assumed the line below dropped the index, but fine to keep it this way if not.